src/renderers/shared/fiber/ReactFiberClassComponent.js

```diff
@@ -103,6 +103,8 @@ module.exports = function(
 
     const instance = workInProgress.stateNode;
     if (typeof instance.shouldComponentUpdate === 'function') {
+      instance.props = workInProgress.memoizedProps;
+      instance.state = workInProgress.memoizedState;
       const shouldUpdate = instance.shouldComponentUpdate(newProps, newState, newContext);
```

Comment:
Shouldn't we also do this reset before `componentWillReceiveProps`? and `componentWillUpdate`?

I think this argues for putting the reset at the start of `updateClassInstance`- so the other 2 lifecycle methods aren't called with stale props. Also, this way, even if the user doesn't define `shouldComponentUpdate` the instance attributes will still be synced before `render` is called.